Overlay encription to use unique key for each pair of hosts

aboch opened the pull request:
- Each ipsec tunnel in the swarm will use a unique key

Signed-off-by: Alessandro Boch <aboch@docker.com>

aboch commented:
ping @diogomonica 
As you suggested offline, I used the hmac to construct the key. PTAL. Thanks.

diogomonica commented on drivers/overlay/encryption.go:
```diff
@@ -43,17 +45,39 @@ func (k *key) String() string {
 	reverse int
 }
 
-func (s *spi) String() string {
```
why is this method not on the pointer?

diogomonica commented on drivers/overlay/encryption.go:
```diff
@@ -43,17 +45,39 @@ func (k *key) String() string {
+}
+
+func (h hpk) String() string {
+	return fmt.Sprintf("Key(FWD: %s, REV: %s)",
```
I don't think there is ever a case where we want to print out the keys in plain-text to the logs.

aboch replied:
I am actually printing the first 5 characters of the key in the debug logs. To help the debugging.

diogomonica replied:
I don't think we should have code in our code base that prints the key. That is asking for trouble.

diogomonica commented on drivers/overlay/encryption.go:
```diff
@@ -43,17 +45,39 @@ func (k *key) String() string {
+	keys hpk
+}
+
+func (ee *encrElem) String() string {
```
So now this one is a method on the pointer? Also, still think we shouldn't be printing plain-text keys: ever.

aboch replied:
If we should not ever print them even a truncated key in the debug level logs, then better remove this.
My assumption was that only an admin can access the debug logs, still the full key wouldn't be printed.

diogomonica replied:
No key material of any sort should ever be pretty anywhere.

diogomonica commented on drivers/overlay/encryption.go:
```diff
@@ -43,17 +45,39 @@ func (k *key) String() string {
+	nodes map[string][]*encrElem
 	sync.Mutex
 }
 
```
Aren't these just debug methods? should they really exist?

aboch replied:
Removed them

diogomonica commented on drivers/overlay/encryption.go:
```diff
@@ -66,18 +90,15 @@ func (e *encrMap) String() string {
 
 func (d *driver) checkEncryption(nid string, rIP net.IP, vxlanID uint32, isLocal, add bool) error {
-	logrus.Debugf("checkEncryption(%s, %v, %d, %t)", nid[0:7], rIP, vxlanID, isLocal)
+	logrus.Debugf("checkEncryption(%s, %v, %d, %t, %t)", nid[0:7], rIP, vxlanID, isLocal, add)
```
What is this `add` boolean?

aboch replied:
This is for debugging. An add==true means a container joined this network and vice-versa, therefore the encryption states (install/remove) must be evaluated.

diogomonica commented on drivers/overlay/encryption.go:
```diff
@@ -143,17 +161,23 @@ func setupEncryption(localIP, advIP, remoteIP net.IP, vni uint32, em *encrMap, k
+
 	for i, k := range keys {
-		spis := &spi{buildSPI(advIP, remoteIP, k.tag), buildSPI(remoteIP, advIP, k.tag)}
+		ee := &encrElem{
```
I think what you need is a method that builds SPIs and Keys and returns the whole object, no? instead of computing them individually.

aboch replied:
Yes, that is an option, given spi and key go hand in hand. I did not go for it because I wanted to contain the refactoring.
I can make the change, it is actually kind of verbose now.

diogomonica commented on drivers/overlay/encryption.go:
```diff
@@ -400,22 +424,32 @@ func buildSPI(src, dst net.IP, st uint32) int {
 	binary.BigEndian.PutUint32(salt, uint32(s))
 	return &netlink.XfrmStateAlgo{
 		Name:   "rfc4106(gcm(aes))",
-		Key:    append(k.value, salt...),
```
This is not a salt right?

aboch replied:
The RFC refers to them as salt value

```
AES-GCM-ESP with a 128 bit key
      The KEYMAT requested for each AES-GCM key is 20 octets.  The first
      16 octets are the 128-bit AES key, and the remaining four octets
      are used as the salt value in the nonce.
```

diogomonica replied:
Cool.

diogomonica commented:
This code is pretty complex, a bunch of ifs without any comments. Can we comment those so we can understand what is happening?

aboch commented:
Thanks @diogomonica Removed logs to avoid troubles, added comments and removed redundant structures following your comment. PTAL when you get a chance.

diogomonica commented on drivers/overlay/encryption.go:
```diff
@@ -43,42 +45,35 @@ func (k *key) String() string {
+	return &encrElem{
+		forwardSPI: buildSPI(src, dst, k.tag),
+		reverseSPI: buildSPI(src, dst, k.tag),
+		forwardKey: buildKey(src, dst, k.tag, k.value),
```
You're calling `buildKey(src, dst, k.tag, k.value)` on both `forwardKey` and `reverseKey`. Doesn't this mean both keys will be the same?

aboch replied:
Thanks, they need to be called with reversed IPs.

aboch replied:
Was trying to parallelize PR and testing...
